test/charms/test_secrets/src/charm.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import json
+from typing import Any
+
+import ops
+
+
+class TestSecretsCharm(ops.CharmBase):
+    _stored = ops.StoredState()
+
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+        self._stored.set_default(secret_id=None)
+        framework.observe(self.on.start, self._on_start)
+        framework.observe(self.on['exec'].action, self._on_exec)
+
+    def _on_start(self, event: ops.StartEvent):
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_exec(self, event: ops.ActionEvent):
+        """Action to execute arbitrary Python code in specific context."""
+        assert event.params['code']
+        rv: dict[str, Any] = {}
+
+        rv['_before'] = self._snapshot()
+
+        try:
+            exec(event.params['code'], globals(), locals())  # noqa
```

Comment:
Per voice discussion, I'd much prefer to avoid the super-generic `exec` approach and include specific actions that we need for the tests.